### PR TITLE
httpserver: drop boost (#8023 dependency)

### PR DESCRIPTION
**Pull request by theuni**
This is a prerequisite for #8023. The httpserver depends on boost::thread::try_join_for, which has no equivalent in std::thread. Rather than going to the effort of porting try_join_for to our interruptible thread wrapper, it's simpler and more efficient to not rely on it in the first place.

755aa05174e06effd758eeb78c5af9fb465e9611 replaces the try_join_for with a future which is set when the callable function finishes. The flow is otherwise the same. I tested that this works as intended by inserting sleeps to force the event_base_loopbreak.

d3773ca9aeb0d2f12dc0c5a0726778050c8cb455 Fixes an incompatibility between boost and std threads. This was likely to be an issue in the future with boost anyway.

Lastly, with the incompatibilities removed, just go ahead and switch to std::thread since there's need for interruption here.

### CODE REVIEW

File: src/httpserver.cpp [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <signal.h>
+#include <future>
```
Comment (jtimon) [inline src/httpserver.cpp]: I guess I should research <functional> and <future> before claiming trivial ACK....

**Comment by sipa** [comment]
Concept ACK

File: src/httpserver.cpp [hunk 2]
```diff
@@ -189,7 +187,7 @@ static bool ClientAllowed(const CNetAddr& netaddr)
 {
     if (!netaddr.IsValid())
         return false;
-    BOOST_FOREACH (const CSubNet& subnet, rpc_allow_subnets)
+    for(const CSubNet& subnet : rpc_allow_subnets)
```
Comment (dcousens) [inline src/httpserver.cpp]: nit: `s/for(/for (/` for consistency

**Comment by dcousens** [comment]
concept && utACK 7e87033

File: src/httpserver.cpp [hunk 3]
```diff
@@ -482,15 +486,11 @@ void StopHTTPServer()
         // master that appears to be solved, so in the future that solution
         // could be used again (if desirable).
         // (see discussion in https://github.com/bitcoin/bitcoin/pull/6990)
-#if BOOST_VERSION >= 105000
-        if (!threadHTTP.try_join_for(boost::chrono::milliseconds(2000))) {
-#else
-        if (!threadHTTP.timed_join(boost::posix_time::milliseconds(2000))) {
-#endif
+        if (threadResult.valid() && threadResult.wait_for(std::chrono::milliseconds(2000)) == std::future_status::timeout) {
```
Comment (laanwj) [inline src/httpserver.cpp]: Nice solution!

**Comment by laanwj** [comment]
utACK 7e87033

**Pull request merged**